File: flask-backend/routes/faults.py
from flask import Blueprint, request, jsonify
from models.fault_record import FaultRecord
from db import db
from datetime import datetime

faults_bp = Blueprint('faults', __name__)

@faults_bp.route('/faults', methods=['GET'])
def get_fault_records():
    """
    获取故障记录列表
    GET /api/faults?equipment_id=1&supplier_id=1&status=reported
    """
    query = FaultRecord.query
    if 'equipment_id' in request.args:
        query = query.filter_by(equipment_id=int(request.args['equipment_id']))
    if 'supplier_id' in request.args:
        query = query.filter_by(supplier_id=int(request.args['supplier_id']))
    if 'status' in request.args:
        query = query.filter_by(status=request.args['status'])
    records = query.all()
    return jsonify([r.to_dict() for r in records])

@faults_bp.route('/faults/<int:id>', methods=['GET'])
def get_fault_record(id):
    """
    获取单个故障记录
    GET /api/faults/:id
    """
    record = FaultRecord.query.get(id)
    if not record:
        return jsonify({'error': '故障记录不存在'}), 404
    return jsonify(record.to_dict())

@faults_bp.route('/faults', methods=['POST'])
def create_fault_record():
    """
    创建故障记录
    POST /api/faults
    """
    data = request.get_json()
    new_record = FaultRecord(**data)
    db.session.add(new_record)
    db.session.commit()
    return jsonify(new_record.to_dict()), 201

@faults_bp.route('/faults/<int:id>', methods=['PUT'])
def update_fault_record(id):
    """
    更新故障记录
    PUT /api/faults/:id
    """
    record = FaultRecord.query.get(id)
    if not record:
        return jsonify({'error': '故障记录不存在'}), 404
    data = request.get_json()
    for key, value in data.items():
        setattr(record, key, value)
    db.session.commit()
    return jsonify(record.to_dict())

@faults_bp.route('/faults/<int:id>', methods=['DELETE'])
def delete_fault_record(id):
    """
    删除故障记录
    DELETE /api/faults/:id
    """
    record = FaultRecord.query.get(id)
    if not record:
        return jsonify({'error': '故障记录不存在'}), 404
    db.session.delete(record)
    db.session.commit()
    return jsonify({'message': '故障记录删除成功'})

@faults_bp.route('/faults/<int:id>/resolve', methods=['POST'])
def resolve_fault(id):
    """
    处理故障
    POST /api/faults/:id/resolve
    请求参数: {"resolution": "处理方案"}
    """
    record = FaultRecord.query.get(id)
    if not record:
        return jsonify({'error': '故障记录不存在'}), 404
    data = request.get_json()
    record.status = 'resolved'
    record.resolved_at = datetime.now()
    record.resolution = data.get('resolution')
    db.session.commit()
    return jsonify(record.to_dict())

File: backend/src/routes/faults.ts
import express from 'express';
import db, { getNextId } from '../database/connection';

const router = express.Router();

router.get('/', async (req, res) => {
  await db.read();
  let records = db.data?.fault_records || [];
  
  if (req.query.equipment_id) {
    records = records.filter(r => r.equipment_id === parseInt(req.query.equipment_id as string));
  }
  if (req.query.supplier_id) {
    records = records.filter(r => r.supplier_id === parseInt(req.query.supplier_id as string));
  }
  if (req.query.status) {
    records = records.filter(r => r.status === req.query.status);
  }
  
  res.json(records);
});

router.get('/:id', async (req, res) => {
  await db.read();
  const record = db.data?.fault_records.find(r => r.id === parseInt(req.params.id));
  if (!record) {
    res.status(404).json({ error: '故障记录不存在' });
    return;
  }
  res.json(record);
});

router.post('/', async (req, res) => {
  await db.read();
  const newRecord = {
    id: getNextId('fault_records'),
    equipment_id: req.body.equipment_id,
    supplier_id: req.body.supplier_id,
    fault_type: req.body.fault_type,
    description: req.body.description,
    status: req.body.status || 'reported',
    resolved_at: req.body.resolved_at,
    resolution: req.body.resolution,
    created_at: new Date().toISOString()
  };
  db.data?.fault_records.push(newRecord);
  await db.write();
  res.status(201).json(newRecord);
});

router.put('/:id', async (req, res) => {
  await db.read();
  const index = db.data?.fault_records.findIndex(r => r.id === parseInt(req.params.id));
  if (index === -1) {
    res.status(404).json({ error: '故障记录不存在' });
    return;
  }
  db.data!.fault_records[index] = {
    ...db.data!.fault_records[index],
    ...req.body
  };
  await db.write();
  res.json(db.data!.fault_records[index]);
});

router.delete('/:id', async (req, res) => {
  await db.read();
  const index = db.data?.fault_records.findIndex(r => r.id === parseInt(req.params.id));
  if (index === -1) {
    res.status(404).json({ error: '故障记录不存在' });
    return;
  }
  db.data?.fault_records.splice(index, 1);
  await db.write();
  res.json({ message: '故障记录删除成功' });
});

router.post('/:id/resolve', async (req, res) => {
  await db.read();
  const index = db.data?.fault_records.findIndex(r => r.id === parseInt(req.params.id));
  if (index === -1) {
    res.status(404).json({ error: '故障记录不存在' });
    return;
  }
  db.data!.fault_records[index] = {
    ...db.data!.fault_records[index],
    status: 'resolved',
    resolved_at: new Date().toISOString(),
    resolution: req.body.resolution
  };
  await db.write();
  res.json(db.data!.fault_records[index]);
});

export default router;
